XPC Codable support

daniel-grumberg opened the pull request:
<!-- What's in this pull request? -->
This PR introduces Encoder and Decoder APIs for serializing and deserializing Codable types to xpc_object_t.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
[rdar://problem/44391240](rdar://problem/44391240)

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

daniel-grumberg committed:
Initial code dump of XPC Encoder, Decoder and associated containers

daniel-grumberg committed:
Add tests for XPCEncoder and friends, this includes simple tests for empty struct and empty class

daniel-grumberg committed:
Add tests to test encoding for single value types, this seems like it should be supported and is

daniel-grumberg committed:
Add tests for for encoding simple structured types

daniel-grumberg committed:
Add test for encoding structured types that encode objects as single value types

daniel-grumberg committed:
Add test for class who shares encoder with its super class

daniel-grumberg committed:
Fix being able to use the same encoder for multiple related objects as long as they all use the same underlying container

daniel-grumberg committed:
Add test for encoding nullable types

daniel-grumberg committed:
Add test for using superEncoder and superDecoder when decoding a derived class

daniel-grumberg committed:
Add test to check for coding path and failure conditions on error

daniel-grumberg committed:
Add support for EncodingError

daniel-grumberg committed:
Fix Decoder coding path test to reflect correct behavior:wq

daniel-grumberg committed:
Add DecodingError support to XPCDecoder and associated types.

The types of errors being thrown are:
- .dataCorrupted when the xpc_object_t does not match the type of the underlying
container, or if an unkeyed container is too small to fit everything it should
- .typeMismatch, when the type of the value in the underlying xpc_object_t is
not correct when decoding
- .keyNotFound when there is no such key in an xpc dictionary wrapped in a
KeyedDecodingContainer

moiseev commented on test/stdlib/TestXPCEncoder.swift:
```diff
@@ -0,0 +1,552 @@
+    }
+
+    static var testValue: Company {
+        return Company(address: Address.testValue, employees: [Employee.testValue])
```
I would sleep better if it had more than one value in the `employees` array. Just in case...

moiseev commented:
@swift-ci Please smoke test

daniel-grumberg commented on stdlib/public/SDK/XPC/XPCDecoder.swift:
```diff
@@ -0,0 +1,110 @@
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+// -----------------------------------------------------------------------------
+///
```
Ok, in which case we need to modify the website, because I got this "template" from there

daniel-grumberg committed:
Improve access control.

daniel-grumberg committed:
Convert XPCDecodingHelpers to be xpc_object_t instance methods via an internal extension

daniel-grumberg commented on stdlib/public/SDK/XPC/XPCEncoder.swift:
```diff
@@ -0,0 +1,125 @@
+        return xpc_double_create(Double(value))
+    }
+
+    static func encodeString(_ value: String) -> xpc_object_t {
```
`xpc_object_t` is actually a protocol, so I am doing these as a case-less enum...

daniel-grumberg committed:
Maek XPCEncodingHelpers a case-less enum to prevent accidental instantiation

daniel-grumberg committed:
Fix ContainerKind enum casing to match style guide

daniel-grumberg committed:
Contrain integer encoding and decoding to fixed width types

daniel-grumberg commented on stdlib/public/SDK/XPC/XPCDecoder.swift:
```diff
@@ -0,0 +1,110 @@
+        guard xpc_get_type(xpcObject) == XPC_TYPE_UINT64 else {
+            throw DecodingError.typeMismatch(to.self,
+                                             DecodingError.Context(codingPath: codingPath,
+                                                                   debugDescription: "Type mismatch.",
```
It would require to add a dictionary that maps the types to their string representation. I am not sure what the policy on adding runtime data for debugging purposes is, so I am gonna wait on this one.

daniel-grumberg committed:
Let encoding integers that are larger than 64 bit fail naturally.

This however should never happen, as there is currently no way to encode a
larger integer type currently.

daniel-grumberg committed:
Rename getXPCObjectForKey(from:...) to getXPCObject(for:...)

daniel-grumberg committed:
Gardening use String(...) instead of Strnig.init(...)

daniel-grumberg commented on stdlib/public/SDK/XPC/XPCKeyedDecodingContainer.swift:
```diff
@@ -0,0 +1,268 @@
+    }
+
+
+    public func decode(_ type: Int.Type, forKey key: Key) throws -> Int {
```
The problem with making a generic thing is that:
```
func decode<I: SignedInteger & FixedWidthInteger>(_ type: S.Type, forKey key: Key) throws -> I
```
conflicts with this:
```
func decode<T: Decodable>(_ type: T.Type, forKey key: Key) throws -> Int
```
when the compiler tries to figure out which definition to use... This is why I originally had to delegate to the generic decodeIntegerType

daniel-grumberg commented on stdlib/public/SDK/XPC/XPCUnkeyedDecodingContainer.swift:
```diff
@@ -0,0 +1,254 @@
+        }
+    }
+
+    public var count: Int?
```
In this case, it is non-sensical. However, the protocol requires it to be optional, where a value of `nil` indicates, that the number of elements is unknown.

daniel-grumberg committed:
Make count of UnkeyedDecodingContainer a computed propoerty

daniel-grumberg committed:
Replace SingleValueEncodingContainerInsertion protocol and associated type, with closures

daniel-grumberg committed:
Gardening generic constraints

daniel-grumberg committed:
Remove unnecessary equality function definitions in testing helper types

daniel-grumberg committed:
Gardening tests

daniel-grumberg committed:
Replace file header doc comments with regular ones

daniel-grumberg commented:
OK as discussed offline, the separate package can be found [here](https://github.com/daniel-grumberg/CodableXPC). If everyone is happy with the state of things we can close this PR.
